File: Neetcode/MaxPathSumBT.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [float("-inf")]
        def dfs(root):
            if root is None:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            pathsum = root.val + max(left, right, 0)
            print(root.val, pathsum)
            res[0] = max([res[0], pathsum, root.val+left+right, root.val])
            return pathsum
        dfs(root)
        return res[0]

File: Neetcode/MaxPathSumBT.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = -1001;
    int bt(TreeNode* node){
        if(node == nullptr){
            return 0;
        }
        int left = bt(node->left);
        int right = bt(node->right);
        res = max({res, node->val, node->val + max(left, right), node->val + left + right});
        return max({0, node->val, node->val + max(left, right)});
    }
    int maxPathSum(TreeNode* root) {
        bt(root);
        return res;
    }
};
